### PR TITLE
[fix] repo level lock

### LINKED ISSUE #1299: SQLite issue when running many jobs in parallel
## 🐛 Bug

When many jobs are launched at exactly the same time it seems there is an issue with sql database access.

### To reproduce

Use slurm to launch ~50 jobs

### Additional context

```Traceback (most recent call last):
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1802, in _execute_context
    self.dialect.do_execute(
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: unable to open database file
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/home/tthonet/src/RL4REC/pretrain_ranker.py", line 77, in <module>
    aim_logger.log_hyperparams(logger_arg_dict)
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/pytorch_lightning/utilities/distributed.py", line 50, in wrapped_fn
    return fn(*args, **kwargs)
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/aim/sdk/adapters/pytorch_lightning.py", line 74, in log_hyperparams
    hparams = self.experiment.meta_run_attrs_tree.subtree('hparams')
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/pytorch_lightning/loggers/base.py", line 43, in experiment
    return get_experiment() or DummyExperiment()
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/pytorch_lightning/utilities/distributed.py", line 50, in wrapped_fn
    return fn(*args, **kwargs)
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/pytorch_lightning/loggers/base.py", line 41, in get_experiment
    return fn(self)
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/aim/sdk/adapters/pytorch_lightning.py", line 53, in experiment
    self._run = Run(
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/aim/sdk/run.py", line 298, in __init__
    self.props
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/aim/sdk/run.py", line 479, in props
    self._init_props()
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/aim/sdk/run.py", line 493, in _init_props
    self._props.experiment = 'default'
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/aim/storage/structured/sql_engine/entities.py", line 116, in experiment
    session.flush()
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3363, in flush
    self._flush(objects)
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3503, in _flush
    transaction.rollback(_capture_exception=True)
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3463, in _flush
    flush_context.execute()
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
    rec.execute(self)
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 236, in save_obj
    _emit_update_statements(
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 1000, in _emit_update_statements
    c = connection._execute_20(
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1614, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 325, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1481, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1845, in _execute_context
    self._handle_dbapi_exception(
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2026, in _handle_dbapi_exception
    util.raise_(
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1802, in _execute_context
    self.dialect.do_execute(
  File "/home/tthonet/src/RL4REC/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) unable to open database file
[SQL: UPDATE run SET updated_at=?, experiment_id=? WHERE run.id = ?]
[parameters: ('2022-01-28 17:17:37.399610', 1, 46)]
(Background on this error at: https://sqlalche.me/e/14/e3q8)
```

### PR DESCRIPTION
**Pull request by devfox-se**
Implemented repo level lock feature

### CODE REVIEW

**Comment by devfox-se** [comment]
This also resolves #1299

**Review by mihran113 — APPROVED** [review]
lgtm

File: aim/sdk/repo.py [hunk 1]
```diff
@@ -109,6 +110,17 @@ def __init__(self, path: str, *, read_only: bool = None, init: bool = False):
...
 
+        self.container_pool: Dict[ContainerConfig, Container] = WeakValueDictionary()
+        self.persistent_pool: Dict[ContainerConfig, Container] = dict()
+        self.container_view_pool: Dict[ContainerConfig, Container] = WeakValueDictionary()
+
+        self._run_props_cache_hint = None
+        self._encryption_key = None
+        self.structured_db = None
+
+        self._lock_path = os.path.join(self.path, '.repo_lock')
```
Comment (alberttorosyan) [inline aim/sdk/repo.py:121]: no need to create lock for `Repo` with remote tracking.

File: aim/sdk/run.py [hunk 2]
```diff
@@ -454,7 +454,8 @@ def _track_impl(
     @property
     def props(self):
         if self._props is None:
-            self._props = self.repo.request_props(self.hash, self.read_only)
+            with self.repo.lock():
```
Comment (alberttorosyan) [inline aim/sdk/run.py:457]: check if the `Repo` is for remote tracking. no need to create a lock for that case.

File: aim/sdk/repo.py [hunk 3]
```diff
@@ -118,19 +130,16 @@ def __init__(self, path: str, *, read_only: bool = None, init: bool = False):
...
-                self.structured_db.run_upgrades()
+        self._lock.acquire()
+        try:
+            if not self.is_remote_repo:
+                self.structured_db = DB.from_path(self.path)
+                if init:
+                    self.structured_db.run_upgrades()
 
-        self._resources = RepoAutoClean(self)
+            self._resources = RepoAutoClean(self)
```
Comment (alberttorosyan) [inline aim/sdk/repo.py:140]: can't we move this out of the `try/except`?

File: aim/sdk/repo.py [hunk 4]
```diff
@@ -118,19 +130,16 @@ def __init__(self, path: str, *, read_only: bool = None, init: bool = False):
...
-        self.container_view_pool: Dict[ContainerConfig, Container] = WeakValueDictionary()
-
-        self._run_props_cache_hint = None
-        self._encryption_key = None
-        self.structured_db = None
-        if not self.is_remote_repo:
-            self.structured_db = DB.from_path(self.path)
-            if init:
-                self.structured_db.run_upgrades()
+        self._lock.acquire()
```
Comment (alberttorosyan) [inline aim/sdk/repo.py:133]: why not use `with self.lock`?

**Review by alberttorosyan — CHANGES_REQUESTED** [review]

File: aim/sdk/repo.py (hunk 1 above)
Comment (devfox-se) [reply aim/sdk/repo.py:121]: good point

**Comment by devfox-se** [comment]
@alberttorosyan all the comments has been addressed, thanks!

**Review by alberttorosyan — APPROVED** [review]
lgtm

**Pull request merged**